framework/src/modules/chain/bft/bft.js

```diff
@@ -84,12 +85,62 @@ class BFT extends EventEmitter {
 			this.constants.startingHeight,
 		);
 
-		await this.loadBlocks({
+		await this.loadBlocksFromStorage({
 			fromHeight: loadFromHeight,
 			tillHeight: lastBlockHeight,
 		});
 	}
 
+	/**
+	 * When blocks deleted send those to BFT to update BFT state
+	 *
+	 * @param {Array.<Object>} blocks - List of all blocks
+	 * @return {Promise<void>}
+	 */
+	async onDeleteBlocks(blocks) {
```

Comment:
`onNewBlock` takes single block, but why `onDeleteBlocks` takes array?
Probably it should be consistent

Reply:
`onNewBlock` works when you receive or forge new block. While in fork recovery, we know we may delete multiple blocks, so why I make these different. If you prefer it to be consistent then I will make `onDeleteBlocks` to work with one bock only.

Reply:
Using event might lead to unexpected behavior. BFT module decides the validity of the blocks based on the information stored internally. Since a block is added/deleted atomically and in a synchronous manner and the addition/removal of block header should also be performed synchronously during the processing of the block addition/deletion. Otherwise, if we rely on events, we may verify an invalid block header in bft because some blockheader was not yet added to bft module through events.

Reply:
Also, I also think this function should be called after each block deletoin.

Reply:
Can't understand your comment, what do you meant by automatically? 

> Since a block is added/deleted atomically 

Also we can't make it synchronous as we have to perform IO call to load blocks.

Reply:
*Atomically. as part of block processing. Sorry for the confusion.

Reply:
Block processing it self is an asynchronous task, so its not necessary to make this one synchronous. And we really can't make it synchronous as we have to load the blocks from storage as well.

Reply:
Maybe synchronous is an incorrect word. What I mean is that each block is processed in a series and after processing one block, the blockchain state is valid; As we process one block, we apply all the state relevant for that block as part of processing i.e, applying transaction, saving transactions and blocks. The same thing happens for block deletion. And the BFT module should behave in a similar fashion. A block header should be added/removed from the BFT module as part of a single block addition/deletion rather than deleting a set of blocks and then removing those headers from the BFT module.
Blocks and transactions are also applied asynchronously, but they are executed within an atomic block processing task, same should go for BFT header addition/deletion.

Reply:
I explained in very detail to @shuse2 why I designed it this way. We know a clear use case of fork elimination which requires deleting multiple blocks, which is obvious and will happen often. And since deleting blocks headers from BFT is a compute intensive task so we should not perform it every block, rather collective. 

If you guys still we should do it for single block, then we can use `[block]` as parameter, by keeping the same implementation. So we can decide while integrating BFT which approach to use.